### PR TITLE
API: sample balancer

### PR DESCRIPTION
**Pull request by joerg-schneider**
Adds a lean implementation of under-/oversampling with `facet.data.SampleBalancer`

### Updates to `Sample` class functionality:

* **Renamed and added methods**:
  - Renamed `Sample.subsample()` to `Sample.resample()` for clarity and updated all references across the codebase. (`[[1]](diffhunk://#diff-9b07fd28838e759e05bdaa5293886d860374fe291bdc6cc43c20f951fc77eab4R87-R89)`, `[[2]](diffhunk://#diff-949d2c2afe9ecca82118a18a596d6b96add302cbe58236f3ed8f1fc175ee1127L721-R721)`, `[[3]](diffhunk://#diff-6bf53585eb2004c8c72920ba5656ede1bce18b35eb2196dce89c1611ff19bc06L295-R295)`, `[[4]](diffhunk://#diff-1259ffd92cd5d66351feaeb48c14087c75c5ebc14d2ba6a2f8d48b3d5fa8f073L133-R133)`, `[[5]](diffhunk://#diff-e40c9c60547e651fa1ec846cdf69bad43750d2d85ad3205ca4b1e8649c5fc46dL45-R45)`)
  - Added a new method, `Sample.rebalance()`, which allows creating a sample with updated weights or adding a new weight column if none exists. (`[[1]](diffhunk://#diff-9b07fd28838e759e05bdaa5293886d860374fe291bdc6cc43c20f951fc77eab4R87-R89)`, `[[2]](diffhunk://#diff-73f5d7068aaeafee9c92db3605924fbf64a25eb2eee9f464e9e04d4315a3746bR271-R345)`)

* **New weight column**:
  - Introduced `COL_WEIGHT` as the default name for the weight column in the `Sample` class. (`[src/facet/data/_sample.pyR68-R71](diffhunk://#diff-73f5d7068aaeafee9c92db3605924fbf64a25eb2eee9f464e9e04d4315a3746bR68-R71)`)
  - Enhanced the `Sample` constructor to support weight column management. (`[src/facet/data/_sample.pyL77-R81](diffhunk://#diff-73f5d7068aaeafee9c92db3605924fbf64a25eb2eee9f464e9e04d4315a3746bL77-R81)`)

### Test suite enhancements:

* **New tests for sample balancing**:
  - Added a comprehensive test suite in `test_sample_balancer.py` to validate the behavior of `SampleBalancer` and its subclasses, including edge cases for undersampling, oversampling, and mixed balancing strategies. (`[test/test/facet/test_sample_balancer.pyR1-R294](diffhunk://#diff-8d21bef31cf69a133bc2dcf9b3759c77f0771baf74b4fa4e2dcdf196c4c0ff16R1-R294)`)

* **Updated existing tests**:
  - Refactored tests to use the renamed `Sample.resample()` method. (`[[1]](diffhunk://#diff-6bf53585eb2004c8c72920ba5656ede1bce18b35eb2196dce89c1611ff19bc06L295-R295)`, `[[2]](diffhunk://#diff-1259ffd92cd5d66351feaeb48c14087c75c5ebc14d2ba6a2f8d48b3d5fa8f073L133-R133)`, `[[3]](diffhunk://#diff-e40c9c60547e651fa1ec846cdf69bad43750d2d85ad3205ca4b1e8649c5fc46dL45-R45)`)

### Codebase maintenance:

* **Imports and typing**:
  - Updated imports in `src/facet/data/__init__.py` to include `_balance`. (`[src/facet/data/__init__.pyR5](diffhunk://#diff-17844c04c6fba92f768d3d79e256582d352b4241e98f9c4f09ffd602e5eb401cR5)`)
  - Added `cast` from `typing` to improve type annotations in `Sample.rebalance()`. (`[src/facet/data/_sample.pyL10-R10](diffhunk://#diff-73f5d7068aaeafee9c92db3605924fbf64a25eb2eee9f464e9e04d4315a3746bL10-R10)`)

These changes improve the clarity, functionality, and test coverage of the `Sample` class and its associated features, ensuring a more robust and user-friendly API.

### CODE REVIEW

**Comment by joerg-schneider** [comment]
PR still work in progress – due for 1.2.x

File: src/facet/data/_balance.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
+from typing import Any, Dict, Iterable, Union
+
+import numpy as np
+import pandas as pd
+
+from pytools.api import AllTracker
+
+from facet.data._sample import Sample
```
Comment (j-ittner) [inline src/facet/data/_balance.py]: `from ._sample import Sample`

**Review by mgelsm — DISMISSED** [review]

**Comment by mgelsm** [comment]
So it should be merged into `1.2.0` instead of `1.1.x` ? Currently set for `1.1.x`

Moreover, don't hesitate to reach out if you want a hand for the documentation or inserting it in a tutorial

**Comment by j-ittner** [comment]
Let's do a final review once this PR has left "draft" state.

**Comment by joerg-schneider** [comment]
As of Monday, updated to new simplified API and offered functionalities – pending some extra refinements, tests (for uniform resampling) and docs/tutorial additions

**Comment by j-ittner** [comment]
We will not implement sample rebalancing, as in most cases sample weighting yields better results.
